### PR TITLE
Correct HTML and JavaScript encoding of `<link>` and `<script>` attribute values

### PR DESCRIPTION
**Pull request by dougbu**
- #4083
- `<link>` tag helper did not HTML encode `href` values in fallback elements
- `<script>` tag helper did not correctly encode any attribute value in fallback elements
 - e.g. double quotes in literal strings would slip through
- only needed to change one existing unit test (!!); so added a bunch

nit: use `Process()`, not `ProcessAsync()` in `<script>` tag helper tests

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs [hunk 1]
```diff
@@ -324,6 +342,37 @@ private void BuildFallbackBlock(TagHelperAttributeList attributes, DefaultTagHel
...
+                stringValue = htmlEncodedString.Value;
+                if (stringValue.Contains("\""))
+                {
+                    stringValue = stringValue.Replace("\"", "&quot;");
+                }
+            }
+            else
+            {
+                var writer = StringWriter;
+                RazorPage.WriteTo(writer, HtmlEncoder, value);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: This `static` method isn't fitting so well in `RazorPage` but I haven't got a better place to throw it. Will leave as-is unless someone has a good suggestion.

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/LinkTagHelperTest.cs [hunk 2]
```diff
@@ -634,24 +638,41 @@ public void RendersLinkTagsForGlobbedHrefResults()
             // Assert
             Assert.Equal("link", output.TagName);
             Assert.Equal("/css/site.css", output.Attributes["href"].Value);
-            Assert.Equal("<link rel=\"stylesheet\" href=\"HtmlEncode[[/base.css]]\" />", output.PostElement.GetContent());
+            var content = HtmlContentUtilities.HtmlContentToString(output, new HtmlTestEncoder());
+            Assert.Equal(expectedContent, content);
         }
 
         [Fact]
-        public void RendersLinkTagsForGlobbedHrefResults_UsingProvidedEncoder()
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/LinkTagHelperTest.cs]: This test was a dupe of `RendersLinkTagsForGlobbedHrefResults()`. Fairly different now. Same deal for `RendersScriptTagsForGlobbedSrcResults_UsesProvidedEncoder()` and `RendersScriptTagsForGlobbedSrcResults()`.

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/LinkTagHelperTest.cs [hunk 3]
```diff
@@ -761,7 +782,143 @@ public void RendersLinkTags_AddsFileVersion_WithRequestPathBase()
...
+        {
+            // Arrange
+            var expectedPostElement = Environment.NewLine +
+                "<meta name=\"x-stylesheet-fallback-test\" content=\"\" class=\"hidden\" />" +
+                "<script>!function(a,b,c){var d,e=document,f=e.getElementsByTagName(\"SCRIPT\")," +
+                "g=f[f.length-1].previousElementSibling," +
+                "h=e.defaultView&&e.defaultView.getComputedStyle?e.defaultView.getComputedStyle(g):g.currentStyle;" +
+                "if(h&&h[a]!==b)for(d=0;d<c.length;d++)e.write('<link rel=\"stylesheet\" href=\"'+c[d]+'\"/>')}(" +
+                "\"JavaScriptEncode[[visibility]]\",\"JavaScriptEncode[[hidden]]\"," +
+                "[\"JavaScriptEncode[[HtmlEncode[[/fallback.css?v=f4OxZX_x_FO5LcGBSKHWXfwtSx-j1ncoSt3SABJtkGk]]]]\"]);" +
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/LinkTagHelperTest.cs]: #4084 covers lack of unbound attributes in the fallback `<link>` element(s).

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/LinkTagHelperTest.cs [hunk 4]
```diff
@@ -761,7 +782,143 @@ public void RendersLinkTags_AddsFileVersion_WithRequestPathBase()
         }
 
         [Fact]
-        public void RendersLinkTags_GlobbedHref_AddsFileVersion()
+        public void RenderLinkTags_FallbackHref_WithFileVersion()
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/LinkTagHelperTest.cs]: Added because we had this test for the `<script>` tag helper but not here.

**Comment by ryanbrandenburg** [comment]
:shipit:, seems straighforward.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs [hunk 5]
```diff
@@ -324,6 +342,37 @@ private void BuildFallbackBlock(TagHelperAttributeList attributes, DefaultTagHel
...
+        private string GetAttributeValue(object value)
+        {
+            string stringValue;
+            var htmlEncodedString = value as HtmlEncodedString;
+            if (htmlEncodedString != null)
+            {
+                // Value likely came from an HTML context in the .cshtml file but may still contain double quotes
+                // since attribute could have been enclosed in single quotes.
+                stringValue = htmlEncodedString.Value;
+                if (stringValue.Contains("\""))
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: This check shouldn't be required. Calling `stringValue.Replace` would no-op if it doesn't have a quote.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs [hunk 6]
```diff
@@ -324,6 +342,37 @@ private void BuildFallbackBlock(TagHelperAttributeList attributes, DefaultTagHel
...
+                {
+                    stringValue = stringValue.Replace("\"", "&quot;");
+                }
+            }
+            else
+            {
+                var writer = StringWriter;
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: Why not use the property directly?

**Comment by pranavkm** [comment]
:shipit:

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs (hunk 5 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: :ok_hand: double-checked the [native source](https://github.com/dotnet/coreclr/blob/f007485a78024b1953a7fe1ffa845be28e93ab6c/src/classlibnative/bcltype/stringnative.cpp#L836) and confirmed that.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs (hunk 6 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/ScriptTagHelper.cs]: :see_no_evil:

**Comment by dougbu** [comment]
ffe2d26

**Pull request closed without merging**